### PR TITLE
add pad transforms with unit tests for lazy resampling

### PR DESCRIPTION
**Pull request by KumoLiu**
Part of #5991.
### Description

This PR adds the pad transforms from https://github.com/Project-MONAI/MONAI/pull/5860 relate to lazy resampling, and it belongs to the second part mentioned in https://github.com/Project-MONAI/MONAI/issues/5991

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/transforms/croppad/array.py [hunk 1]
```diff
@@ -142,6 +143,27 @@ def _pt_pad(img: torch.Tensor, pad_width, mode, **kwargs) -> torch.Tensor:
         # torch.pad expects `[B, C, H, W, [D]]` shape
         return pad_pt(img.unsqueeze(0), pt_pad_width, mode=mode, **kwargs).squeeze(0)
 
+    @staticmethod
+    def pad_nd(img_t, to_pad_, mode, **kwargs):
```
Comment (wyli) [inline monai/transforms/croppad/array.py:147]: could you please remove the underscore `_`,  `to_pad_` -> `to_pad` and also add some docstring? I can also push some docstring to your PR directly before merging this PR.

Comment (KumoLiu) [reply monai/transforms/croppad/array.py:147]: OK, will do that.

File: tests/padders.py [hunk 2]
```diff
@@ -109,3 +112,27 @@ def pad_test_kwargs(self, unchanged_slices, **input_param):
                             inv = padder.inverse(result)
                             assert_allclose(im, inv, type_test=False)
                             self.assertEqual(inv.applied_operations, [])
+
+    def pad_test_pending_ops(self, input_param, input_shape):
```
Comment (wyli) [inline tests/padders.py:116]: would be great to have a test about lazily combining multiple crop/pad operations chained in different ordering will give the expected consistent outcomes.

File: monai/transforms/croppad/functional.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        affine=xform,
+        extra_info=extra_info,
+        orig_size=img_size,
+        transform_info=transform_info,
+        lazy_evaluation=transform_info.get(TraceKeys.LAZY_EVALUATION, False),
+    )
+    out = convert_to_tensor(img.as_tensor() if isinstance(img, MetaTensor) else img, track_meta=get_track_meta())
+    if transform_info.get(TraceKeys.LAZY_EVALUATION, False):
+        return out.copy_meta_from(meta_info) if isinstance(out, MetaTensor) else meta_info
+    out = monai.transforms.Pad.pad_nd(out, to_pad_, mode, **kwargs) if do_pad else out
```
Comment (Nic-Ma) [inline monai/transforms/croppad/functional.py:58]: I think `functional.py` should be the lower level operations of transforms, but here `pad_func` calls `Pad`'s static method, then `Pad` also calls `pad_func`, I think we should avoid this `A` <-> `B` logic, can we move the `pad_nd` function of `Pad` to here directly?

Thanks.

File: monai/transforms/spatial/functional.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    meta_info = TraceableTransform.track_transform_meta(
+        img,
+        sp_size=spatial_size,
+        affine=None if affine_unchanged and not lazy_evaluation else xform,
+        extra_info=extra_info,
+        orig_size=original_spatial_shape,
+        transform_info=transform_info,
+        lazy_evaluation=lazy_evaluation,
+    )
+    img = img.as_tensor() if isinstance(img, MetaTensor) else img
```
Comment (Nic-Ma) [inline monai/transforms/spatial/functional.py:98]: Could you please help me understand why we convert `MetaTensor` to `Tensor` here?
If must, can we execute it after line 102, then no need to convert it back to `MetaTensor`?

Thanks.

File: tests/test_spatial_resample.py [hunk 5]
```diff
@@ -142,7 +142,7 @@ def test_flips(self, img, device, data_param, expected_output):
                 img = img.to(device)
             out = SpatialResample()(img=img, **data_param)
             assert_allclose(out, expected_output, rtol=1e-2, atol=1e-2)
-            assert_allclose(out.affine, data_param["dst_affine"])
```
Comment (Nic-Ma) [inline tests/test_spatial_resample.py:145]: May I know why remove the verification for predefined `data_param["dst_affine"]`?

Thanks.

Comment (KumoLiu) [reply tests/test_spatial_resample.py:145]: Hi @Nic-Ma, we don't remove this verification. We added `to_affine_nd` to make the dimension right. If the data is not metatensor, we will add a fake affine.

File: monai/transforms/croppad/functional.py (hunk 3 above)
Comment (KumoLiu) [reply monai/transforms/croppad/functional.py:58]: Hi Nic, if like that, we may also need to move `_np_pad` and `_pt_pad` since `pad_nd` also calls these two functions. What do you think? @wyli

File: monai/transforms/spatial/array.py [hunk 6]
```diff
@@ -340,14 +229,6 @@ class ResampleToMatch(SpatialResample):
     """Resample an image to match given metadata. The affine matrix will be aligned,
     and the size of the output image will match."""
 
-    def update_meta(self, img: torch.Tensor, dst_affine=None, img_dst=None):
-        if dst_affine is not None:
-            super().update_meta(img, dst_affine)
```
[the remark this answers is not in the document]
Comment (KumoLiu) [reply monai/transforms/spatial/array.py:345]: Hi @Nic-Ma, do you mean save meta information in a dictionary like `meta_dict`? I think the same logic is retained here as in the original.

File: monai/transforms/croppad/functional.py (hunk 3 above)
Comment (wyli) [reply monai/transforms/croppad/functional.py]: Sure, I don't have a preference here...

Comment (KumoLiu) [reply monai/transforms/croppad/functional.py]: Ok, will move these functions, and it will also avoid the duplicated doc strings.

File: monai/transforms/spatial/functional.py (hunk 4 above)
Comment (wyli) [reply monai/transforms/spatial/functional.py:98]: I think line 102 is a shallow copy, so we should drop the current meta first. Skipping this will cause some bugs.

File: monai/transforms/croppad/functional.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+            (PyTorch) {``"constant"``, ``"reflect"``, ``"replicate"``, ``"circular"``}.
+            One of the listed string values or a user supplied function. Defaults to ``"constant"``.
+            See also: https://numpy.org/doc/1.18/reference/generated/numpy.pad.html
+            https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html
+        transform_info: a dictionary with the relevant information pertaining to an applied transform.
+        kwargs: other arguments for the `np.pad` or `torch.pad` function.
+            note that `np.pad` treats channel dimension as the first dimension.
+    """
+
+    def _np_pad(img: torch.Tensor, pad_width: list[tuple[int, int]], mode: str, **kwargs) -> torch.Tensor:
```
Comment (wyli) [inline monai/transforms/croppad/functional.py:48]: This makes the functions not reachable by the users, is it really necessary to hide? Ateast I see pad_nd is a good utility. Previously the user can call Pad.pad_nd

Comment (KumoLiu) [reply monai/transforms/croppad/functional.py:48]: Can we set default value for `transform_info` and `kwargs` like this `, transform_info: dict = {}, kwargs: dict = {}`? Then users could directly call `pad_func`.

Comment (wyli) [reply monai/transforms/croppad/functional.py:48]: I don't think they are the same thing, the utilities can keep the data types, right?

Comment (KumoLiu) [reply monai/transforms/croppad/functional.py:48]: Oh, I see.

Comment (KumoLiu) [reply monai/transforms/croppad/functional.py:48]: Or we could add a flag in `pad_func`, if true, just return `pad_nd`? I can't find another way to avoid `A` <-> `B` logic.

File: tests/padders.py (hunk 2 above)
Comment (KumoLiu) [reply tests/padders.py:116]: Hi @wyli, since we don't have the compose logic in this PR, do you mean combining multiple operations manually?

File: tests/test_spatial_resample.py [hunk 8]
```diff
@@ -142,7 +142,7 @@ def test_flips(self, img, device, data_param, expected_output):
...
+            assert_allclose(to_affine_nd(len(out.shape) - 1, out.affine), data_param["dst_affine"])
 
     @parameterized.expand(TEST_4_5_D)
     def test_4d_5d(self, new_shape, tile, device, dtype, expected_data):
```
Comment (yiheng-wang-nv) [inline tests/test_spatial_resample.py:148]: Hi @wyli , here I'm a bit confused with this test case. 
According to: https://github.com/Project-MONAI/MONAI/pull/6031/files#diff-3a1b9fb96f9cde878a33605fcbd39cf1af51a79c285fe662d80a00f0c43e2d5fL205
the transform does not support >3d input images, is it correct?

File: monai/transforms/croppad/functional.py (hunk 7 above)
Comment (wyli) [reply monai/transforms/croppad/functional.py:48]: what is A<->B, do you mean `pad_nd` <-> `pad_func`?

File: tests/padders.py (hunk 2 above)
Comment (wyli) [reply tests/padders.py:116]: Yes

File: monai/transforms/croppad/functional.py (hunk 7 above)
Comment (wyli) [reply monai/transforms/croppad/functional.py:48]: We can have a few functions here:


def _np_pad():
  ...

def _pt_pad():
  ...

def pad_nd():
  ...

def pad_func():
  ...

It’s also fine if you want to put these functions elesewhere. But please make pad_nd and pad_func available to the users. pad_nd is a useful low level implementation, pad_func is the primary functional used by the array transform.

File: tests/padders.py (hunk 2 above)
Comment (KumoLiu) [reply tests/padders.py:117]: Hi @wyli, I have added a test about lazily combining multiple pad operations chained in different ordering in this commit. https://github.com/Project-MONAI/MONAI/pull/6031/commits/81de9e2f1b6aa82ad3ff3cb9754609a36a22c537
But here I have one question, I only test multiply pad operations with the same pad mode. If with different pad modes, which mode should I specify to `apply_transforms`?

File: monai/transforms/croppad/functional.py (hunk 7 above)
Comment (KumoLiu) [reply monai/transforms/croppad/functional.py]: Have moved `_np_pad`, `_pt_pad`, and `pad_nd` out of `pad_func` as you suggested!

File: tests/padders.py (hunk 2 above)
Comment (wyli) [reply tests/padders.py:117]: yes, in theory when we lazily evaluate the pending operations we should check the compatibility of the padding modes across the transforms https://github.com/Project-MONAI/MONAI/blob/b9e17e8186cb25c8781e2c82de3b4915081df48b/monai/transforms/lazy/functional.py#L83-L85 the `is_compatible_apply_kwargs` API is defined, for now we haven't implemented any specific rules. In the first version, I don't thinks it's a high-priority feature..

**Review by wyli — APPROVED** [review]

File: monai/transforms/croppad/functional.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    return out
+
+
+def _pt_pad(img: torch.Tensor, pad_width: list[tuple[int, int]], mode: str, **kwargs) -> torch.Tensor:
+    pt_pad_width = [val for sublist in pad_width[1:] for val in sublist[::-1]][::-1]
+    # torch.pad expects `[B, C, H, W, [D]]` shape
+    return pad_pt(img.unsqueeze(0), pt_pad_width, mode=mode, **kwargs).squeeze(0)
+
+
+def pad_nd(img: torch.Tensor, to_pad: list[tuple[int, int]], mode: str, **kwargs):
```
Comment (Nic-Ma) [inline monai/transforms/croppad/functional.py:48]: it seems this `pad_nd` function is only used within this file, should we make it as private? `_pad_nd`.

Thanks.

Comment (KumoLiu) [reply monai/transforms/croppad/functional.py:48]: Hi @Nic-Ma, Wenqi suggested it to be available to the user in this [reply](https://github.com/Project-MONAI/MONAI/pull/6031#discussion_r1115749051).

File: monai/transforms/croppad/functional.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        return _pad(img, pad_width=to_pad, mode=mode, **kwargs)
+    except (ValueError, TypeError, RuntimeError) as err:
+        if isinstance(err, NotImplementedError) or any(
+            k in str(err) for k in ("supported", "unexpected keyword", "implemented")
+        ):
+            return _np_pad(img, pad_width=to_pad, mode=mode, **kwargs)
+        raise ValueError(f"{img.shape} {to_pad} {mode} {kwargs} {img.dtype} {img.device}") from err
+
+
+def pad_func(img: torch.Tensor, to_pad: list[tuple[int, int]], mode: str, transform_info: dict, kwargs: dict):
```
Comment (Nic-Ma) [inline monai/transforms/croppad/functional.py:84]: I think we should use `dict | None = None` for `kwargs` or no typehint, refer to:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/dataset.py#L548

Thanks.

File: monai/transforms/spatial/array.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/spatial/array.py:345]: Hi @KumoLiu ,

I think the original logic will update meta no matter the image is `MetaTensor` or not, but the changed logic of this PR will only update meta when `if isinstance(img, MetaTensor)` as line 280, right?

Thanks.

File: monai/transforms/spatial/functional.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/transforms/spatial/functional.py:98]: Hi @KumoLiu ,

Could you please help add 1 line comment here to explain the code logic? Otherwise, I feel it's very confusing why we convert `MetaTensor` to `Tensor`.

Thanks in advance.

File: monai/transforms/croppad/functional.py (hunk 10 above)
Comment (wyli) [reply monai/transforms/croppad/functional.py:84]: this `kwargs` doesn't seem to support `None`

**Review by Nic-Ma — COMMENTED** [review]
Put some minor comments inline, others look good to me.

Thanks.

File: monai/transforms/spatial/array.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (KumoLiu) [reply monai/transforms/spatial/array.py:345]: Hi @Nic-Ma, the original logic also only update affine when data is `MetaTensor` since we didn't have a meta dict before. If data is `non-MetaTensor`, affine information would be lost. Or did I misunderstand it?
https://github.com/Project-MONAI/MONAI/blob/b9e17e8186cb25c8781e2c82de3b4915081df48b/monai/transforms/spatial/array.py#L189-L190

File: monai/transforms/spatial/functional.py (hunk 4 above)
Comment (KumoLiu) [reply monai/transforms/spatial/functional.py:99]: Have added comments in this [commit](https://github.com/Project-MONAI/MONAI/pull/6031/commits/4281a597be03066743ac19577baabaf944f86d5c). Thanks!

File: monai/transforms/croppad/functional.py (hunk 10 above)
Comment (KumoLiu) [reply monai/transforms/croppad/functional.py]: Have removed typehint for `kwargs`, thanks!

File: monai/transforms/spatial/array.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/transforms/spatial/array.py:345]: OK, seems we have bug in the dev, if input is not MetaTensor, it still updates meta.
Thanks for the fix!

File: monai/transforms/croppad/functional.py (hunk 9 above)
Comment (Nic-Ma) [reply monai/transforms/croppad/functional.py:48]: OK, I see.

Thanks.

**Review by Nic-Ma — APPROVED** [review]
Thanks for the quick update.
Looks good to me now.

File: monai/transforms/croppad/functional.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    return pad_pt(img.unsqueeze(0), pt_pad_width, mode=mode, **kwargs).squeeze(0)
+
+
+def pad_nd(img: torch.Tensor, to_pad: list[tuple[int, int]], mode: str, **kwargs):
+    """
+    PyTorch/Numpy pad ``img`` with integers ``to_pad`` amounts. Depending on the ``mode`` and input dtype,
+    a suitable backend will be used automatically.
+
+    Args:
+        img: data to be transformed, assuming `img` is channel-first and padding doesn't apply to the channel dim.
```
Comment (wyli) [inline monai/transforms/croppad/functional.py:54]: the docstring of `pad_nd` is not correct here, the `to_pad` parameter has to include the amounts along the channel dimension as well.

Comment (KumoLiu) [reply monai/transforms/croppad/functional.py:54]: Thanks for pointing it out, I will update it in the later pr.

Comment (KumoLiu) [reply monai/transforms/croppad/functional.py:54]: updated docstring in this [commit](https://github.com/Project-MONAI/MONAI/pull/6068/commits/e13de51ed80c0634a2e6895f65d61d92b80deddf).
